_posts/2016-06-24-scripting-veritas-netbackup.md

```diff
@@ -0,0 +1,256 @@
+image:      veritas_bash.png
+category:   Deployment
+tags:       linux, unix, scripting, bash, vmware, netbackup, policy, slp, replication, air
+---
+
+Veritas NetBackup is a powerful backup platform, with extensive range of functionality. Some of the key features are the ability to backup a VMware Virtual Machine (VM) and to replicate backup images to remote sites.
+
+Our requirement is to have a repeatable process that can be used to create many tens, even hundreds of policies to backup VMs and replicate them to one or two remote sites. The use of the NetBackup Java Remote Administration Console (RAC) to perform this configuration would be very time consuming and the risk of human error high. The consequences of this could mean the inability to restore data in the event of a disaster.
+
+# The Challenge
+
+NetBackup does not (yet) have any APIs that can be used to programatically configure it, but it does come with extensive suite of CLI commands that can be used to configure most aspects of the product. However, documentation and examples of the use of these commands can be sparse and difficult to understand.
+
+Some commands can be simple with a few, single option switches, but some can be very complex with many key-value pairs per switch, or with a variable number of key-value pairs depending on how many sub-operations you are configuring within a given object.
+
+A policy can be as simple as a simple filesystem backup to local Storage Units (STU) with one schedule, or can be very complex with the backup of multiple VMs with a query language, multiple schedules, with multiple Storage Lifecycle Policies (SLP) each performing Auto Image Replication (AIR) to multiple sites, each of which require import policies to have been created on remote NetBackup Domains.
+
+Putting all this together in a single one-liner tool with a few custom options requires that many CLI commands need to interact with the output from a previous command and modify their command line switches and options accordingly.
+
+I'm not going to go into the meaning of every option for every command, but simply present the commands that worked for us. If you require more details, please drop me a line or consult the man pages or Veritas Tech Support.
+
+### Naming Conventions
+
+This is a good point to discuss naming conventions. In order to be able to maintain consistency across multiple policy creations, and especially across multiple users, it is essential to have a defined naming convention. For our purposes, the following naming convention was adopted, in which the various components of NetBackup are broken down into segments according to the Policy Type:
+
+* All text is to be in lower case (to avoid complications in a mixed UNIX/Windows environment).
+* Only alphanumeric characters and underscore are permitted.
+* Underscores should be used in place of spaces.
+* Any segment in <angled_brackets> is mandatory.
```

Comment:
Can you put `<angle_brackets>` (and `[square_brackets]`) in backticks as it's coming through as HTML :-(